### DISCUSSION TITLE
How to do a HTTP request when override tryResolveKey method of FileNameResolver?

### DISCUSSION THREAD

**Discussion by Snailya** [iTwin/itwinjs-core#4312] [Q&A]
When open a imodel through RPC, I want to pass in only the id of the snapshot, and resolve the id to filename through FileNameResolver. What I encounter is that the conversion from id to filename is done by an external HTTP request which is asynchronous, but the tryResolveKey method is synchronous. Is it able to call a asynchorous method in a synchronous method and return the value after the asynchourous method resolved? Could you give me a sample code?

**Comment by kabentley** [comment]
you can find an example of a `FileNameResolver` [here](https://github.com/iTwin/itwinjs-core/blob/f6a992334a6c7802524ecf132f1531d37cda5a2f/full-stack-tests/core/src/backend/backend.ts#L134-L149)

**Reply by pmconne** [reply]
The problem appears to be that `FileNameResolver`'s methods are all synchronous.
@Snailya why do you need to make a http request to resolve a filename on a local file system?

**Reply by Snailya** [reply]
As I have an external file manage system that manages all types of files, and I don't want to expose the actual storage location for the end user.

**Reply by kabentley** [reply]
that's what FileNameResolver is for (see example). But why does your implementation of it need to be asynchronous?

**Reply by Snailya** [reply]
As our file is host by another file system that so I should solve the actual file location through http request.

**Reply by kabentley** [reply]
The string returned by FileNameResolver must be a path to a local file that can be opened with "fopen".  Are you attempting to use a network drive? If so, i don't understand why there'd be a need for an http request to get the filename. 

I suppose we could change the tryResolveMethod to also support  string | `Promise<string>`, but frankly i can't understand your use case. What does your http request actually do?
